implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake_worker.rs

```diff
@@ -0,0 +1,344 @@
+        };
+
+        // set the remote route by taking the most up to date message return route
+        // In the case of the initiator the first return route mentions the secure channel listener
+        // address so we need to wait for the return route corresponding to the remote handshake worker
+        // when it has been spawned
+        self.remote_route = Some(message.return_route());
+
+        if let SendMessage(message) = self
+            .state_machine
+            .on_event(ReceivedMessage(Vec::<u8>::decode(
+                &message.into_transport_message().payload,
+            )?))
+            .await?
+        {
+            context
+                .send_from_address(
+                    self.remote_route()?,
+                    message,
+                    self.addresses.decryptor_remote.clone(),
+                )
+                .await?
+        };
+
+        // if we reached the final state we can make a pair of encryptor/decryptor
+        if let Some(final_state) = self.state_machine.get_handshake_results() {
+            // start the encryptor worker and return the decryptor
+            self.decryptor_handler = Some(self.finalize(context, final_state).await?);
+            self.callback_sender.send(()).await?;
+        };
+
+        Ok(())
+    }
+}
+
+impl HandshakeWorker {
+    /// Create a new HandshakeWorker with a role of either INITIATOR or RESPONDER
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) async fn create(
+        context: &Context,
+        secure_channels: Arc<SecureChannels>,
+        addresses: Addresses,
+        identifier: IdentityIdentifier,
+        trust_policy: Arc<dyn TrustPolicy>,
+        decryptor_outgoing_access_control: Arc<dyn OutgoingAccessControl>,
+        credentials: Vec<Credential>,
+        trust_context: Option<TrustContext>,
+        remote_route: Option<Route>,
+        timeout: Option<Duration>,
+        role: Role,
+    ) -> Result<()> {
+        let vault = to_xx_vault(secure_channels.vault());
+        let identities = secure_channels.identities();
+        let state_machine: Box<dyn StateMachine> = if role.is_initiator() {
+            Box::new(
+                InitiatorStateMachine::new(
+                    vault,
+                    identities,
+                    identifier.clone(),
+                    credentials.clone(),
+                    trust_policy.clone(),
+                    trust_context.clone(),
+                )
+                .await?,
+            )
+        } else {
+            Box::new(
+                ResponderStateMachine::new(
+                    vault,
+                    identities,
+                    identifier.clone(),
+                    credentials.clone(),
+                    trust_policy.clone(),
+                    trust_context.clone(),
+                )
+                .await?,
+            )
+        };
+
+        let (mut callback_waiter, callback_sender) = ockam_node::callback::new_callback();
```

Comment:
I would like to attempt to change that to oneshot channel in my following PR

Reply:
Also a little bit concerned with the fact that it's not used for the responder, but it should theoretically be fine. The receiver is just discarded.